Add OTLP HTTP MetricExporter max_export_batch_size

tammy-baylis-swi opened the pull request:
# Description

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Contrib Repo Change?

<!--
Answer the following question based on these examples of changes that would require a Contrib Repo Change:
- [The OTel specification](https://github.com/open-telemetry/opentelemetry-specification) has changed which prompted this PR to update the method interfaces of `opentelemetry-api/` or `opentelemetry-sdk/`
- The method interfaces of `test/util` have changed
- Scripts in `scripts/` that were copied over to the Contrib repo have changed
- Configuration files that were copied over to the Contrib repo have changed (when consistency between repositories is applicable) such as in
    - `pyproject.toml`
    - `isort.cfg`
    - `.flake8`
- When a new `.github/CODEOWNER` is added
- Major changes to project information, such as in:
    - `README.md`
    - `CONTRIBUTING.md`
-->

- [ ] Yes. - Link to PR: 
- [ ] No.

# Checklist:

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

tammy-baylis-swi commented:
I think the [aiohttp-client test failure](https://github.com/open-telemetry/opentelemetry-python/actions/runs/14935684759/job/41962430435?pr=4576) is a hiccup from the recent release, not from changes in this PR.

tammy-baylis-swi commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -219,27 +243,448 @@ def export(
+                    # with different accessors for data points, etc
+                    # We maintain these structures throughout batch calculation
+                    current_data_points = []
+                    if metric.HasField("sum"):
```
Another difference between http protobuf and the gRPC classes are that protobuf does not define a `DataPointT` to encompass all metric types. So I've added some `if-elif` throughout. Not sure if there is a better way.

lzchen replied:
Unfortunate if this is the only way :(

tammy-baylis-swi replied:
Yes, a bit easier to look at. Changed in [acc1c7c](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/acc1c7c6f5746a401253f8f3a8af8b85f25829e3)

tammy-baylis-swi commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -219,27 +243,448 @@ def export(
+                )
+            )
+
+    def _get_split_resource_metrics_pb2(
```
This helper is used to write actual pb2 objects before each `yield` back to the `export` caller.

lzchen commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -219,27 +243,448 @@ def export(
+
+                    if split_resp.ok:
+                        export_result = MetricExportResult.SUCCESS
+                    elif self._retryable(split_resp):
```
Wouldn't we need to remove the success/failure batches from the entire list when we attempt a retry? For example, batch size of 1 and there are 2 metric datas. The first one returns retryable and the second one returns failure. When the loop returns to the outer loop, we don't want to retry the one that failed again right?

tammy-baylis-swi replied:
Good call, no we don't want to retry a non-retryable failure!

I reorganized the unbatched-vs-batched and, for the latter, made `batch` the outer loop and `delay` the inner loop with breaks for clarity in [f1ef6c4](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/f1ef6c404faab53a8eb9685206e8e8010a0b5f77).

tammy-baylis-swi replied:
Hi @lzchen any other changes to suggest?

tammy-baylis-swi commented:
Hi @lzchen and @open-telemetry/python-approvers , please may I have a new review of this PR.

In [25e8be9](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/25e8be93f0d66702492685dfc3770347e8b7253a) I resolved merge conflicts which required some redesign of HTTP metrics batching to match the timeout updates in https://github.com/open-telemetry/opentelemetry-python/pull/4564. That was the main change since last approval; the `split` helpers are the same.

The last Cassandra and Celery instrumentor test failures seem unrelated and are citing commit hashes like similar failures in other PRs.

pmcollins commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -211,33 +235,468 @@ def export(
+        # If no batch size configured, export as single batch with retries as configured
+        if self._max_export_batch_size is None:
+            for retry_num in range(_MAX_RETRYS):
+                resp = self._export(
```
What do you think about consolidating this block with the one on line 277? I think at least the retry logic can be factored out.

tammy-baylis-swi replied:
Yes good idea! Extracted `_export_with_retries` in [d748c12](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/d748c128cc8ea39f5875cfb0824286e4d5072f93)

tammy-baylis-swi commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -211,33 +235,468 @@ def export(
-        return MetricExportResult.FAILURE
+
+    def _get_split_resource_metrics_pb2(
+        self,
```
How is something like [e7c80e5](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/e7c80e5c5a286843079bd30b0d012686099daa95)?

tammy-baylis-swi commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -230,15 +256,282 @@ def export(
+
+        # Else, export in batches of configured size
+        split_metrics_batches = list(self._split_metrics_data(serialized_data))
+        export_result = MetricExportResult.SUCCESS
```
In [dbd639f](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/dbd639f2f6546f3b00aaaf2d64f4e8ae71054bea) I changed the logic to be more like how gRPC metric exporter does it which is to return the last result: https://github.com/open-telemetry/opentelemetry-python/pull/2809/files#diff-0a8e4a5e11e9b47310b6939eb273ec3f3c5347bc9fb8ace8bff245bf5d078c29R272-R275. Is it ok this way or should I still remove this first init of `export_result`?

tammy-baylis-swi replied:
I will keep as is 😃

tammy-baylis-swi commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -230,15 +256,282 @@ def export(
+        # Return SUCCESS only if all batches succeeded
+        return export_result
+
+    def _split_metrics_data(
```
Should be! Moved in [eed52e3](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/eed52e3c51c8d77a05fe1f3bbd13679af5d88b1f)

42Questions commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -259,6 +286,40 @@ def export(
+            export_result = self._export_with_retries(
+                split_metrics_data.SerializeToString(),
+                deadline_sec,
+            )
```
Is it ok for data loss to happen here?
For example if there are two batches and the first one fails, and the first one succeeds.
I can see that there is a test that confirms the behaviour `test_export_retries_with_batching_failure_first`, would just like to make sure that it's intentional.

tammy-baylis-swi replied:
Thank you both. In [6f5efbf](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/6f5efbf520fc076381cfda6f6a1f99669a52e004) I've set it to fail fast instead.

42Questions commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -276,6 +337,396 @@ def force_flush(self, timeout_millis: float = 10_000) -> bool:
         return True
 
 
+def _split_metrics_data(
```
This seems quite verbose, are the if elif chains needed?
I think all metric types (sum, histogram, gauge, etc.) have a data_points field within their data container, could we use WhichOneof() and CopyFrom() to handle them agnostically?

anuraaga commented on exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py:
```diff
@@ -291,6 +352,310 @@ def force_flush(self, timeout_millis: float = 10_000) -> bool:
+def _split_metrics_data(
+    metrics_data: pb2.MetricsData,
+    max_export_batch_size: int | None = None,
+) -> Iterable[pb2.MetricsData]:
```
I think this is also supposed to be `Iterable[ExportMetricsServiceRequest]`

anuraaga reviewed (commented):
Hi @tammy-baylis-swi @lzchen. I was trying to wrap my head around this PR while merging main into my branch, but I am wondering, does it actually work? The types seem to be very off, and I think only work since everything is mocked heavily in the unit tests - the bytes sent will probably be the wrong type based on my skim.

If that sounds right, could it be reverted and tried again?

tammy-baylis-swi commented:
> > Hi @tammy-baylis-swi @lzchen. I was trying to wrap my head around this PR while merging main into my branch, but I am wondering, does it actually work? The types seem to be very off, and I think only work since everything is mocked heavily in the unit tests - the bytes sent will probably be the wrong type based on my skim.
> > If that sounds right, could it be reverted and tried again?
> 
> Hi @anuraaga , I'll take a look. Meanwhile, could you share more about your issue? What specific error do you get when you opt into this feature while instrumenting your test services?

[This gist](https://gist.github.com/tammy-baylis-swi/d19b385742da10de112c1da6be17895f) contains a quick demo app and output using changes from this PR and `max_export_batch_size: 2`. The debug log shows the batching is working as expected where `ResourceMetrics #0` contains `trace.service.counter_a` and `trace.service.counter_b` (first batch) and `ResourceMetrics #1` contains `trace.service.counter_c` (second batch). If there are problems with your use case(s), please could you create a new issue.
